src/scicloj/02_numpy.clj

```diff
@@ -140,7 +155,53 @@ np.random.randint(0, 10, (3, 3))
 np.eye(3)
 ```"]
 
-;; N/A
+["We currently do not have a good approximation of this technique.  However,
+it is easy enough to compute an identity matrix using an outer product operation."]
+
+(defn index
+  "Like range, but for tensors.  Enumerates the indicies."
+  [shape]
+  (let [space  (apply * shape)
+        rshape (reverse shape)
+        muls   (cons 1 (pop (vec (reductions * rshape))))]
+    (map
+     (fn [idx]
+       (reverse
+        (map (fn [p n]
+               (-> idx
+                   (quot n)
+                   (mod p)))
+             rshape
+             muls)))
+     (range space))))
+
+(index [3 3])
+
+(defn outer-product [f a b]
+  (let [final-shape (concat (dtype/shape a) (dtype/shape b))
+        size        (apply * final-shape)
+        index-a     (index (dtype/shape a))
+        index-b     (index (dtype/shape b))
+        indexes     (for [a index-a
+                          b index-b]
+                      [a b])]
+    (-> (dtype/make-reader
+         :object
+         size
+         (let [[a' b'] (nth indexes idx)]
+           (f (apply dtt/mget a a')
+              (apply dtt/mget b b'))))
+        (dtt/->tensor)
+        (dtt/reshape final-shape))))
+
+(defn eye [n]
+  (letfn [(= [a b] (if (clojure.core/= a b) 1 0))]
+    (outer-product = (dtt/->tensor (range n))
+                   (dtt/->tensor (range n)))))
+
+(eye 3)
+
+(eye 5)
 
```

Comment:
```clojure
(defn reduce-axis
  ([reduce-fn tensor axis]
   (let [rank (count (dtype/shape tensor))
         dec-rank (dec rank)
         axis (if (>= axis 0)
                axis
                (+ rank axis))
         shape-idxes (remove #(= axis %) (range rank)0)
         ;;transpose the tensor so the reduction axis is the last one
         tensor (if-not (= dec-rank axis)
                  (dtt/transpose tensor (concat shape-idxes [axis]))
                  tensor)
         ;;slice to produce n sequence of data
         slices (dtt/slice tensor dec-rank)
         result-shape (mapv (dtype/shape tensor) shape-idxes)]
     (-> (dtype/emap reduce-fn
                     :object
                     slices)
         ;;reshape to the result shape
         (dtt/reshape result-shape))))
  ([reduce-fn tensor]
   (reduce-axis reduce-fn tensor -1)))
```

Reply:
WOW!  What a Christmas gift.  :gift: 
That is seriously clever.  

How would you feel about this adjustment:

```clojure
(dtype/emap reduce-fn
                     :object
                     slices)
```
(which requires an API of `(reduce-axis (partial apply +) some-tensor)`
to
```clojure 
(dtype/emap (partial reduce reduce-fn)
                     :object
                     slices)
```
(which requires an API of `(reduce-axis + some-tensor)`?

Reply:
Then things like dfn/mean won't work.  You have to account for the fact that some functions expect aggregates

Reply:
Like, I could have used dfn/sum, dfn/mean, lots of the stats functions could be useful here.

Reply:
Ahhh got it.  That makes sense!

Reply:
I had to temporarily revert to the less performant version of `reduce-axis` because the presented version violates the following constraint:

```text
initial shape        2 3 5
reduce-axis     0      3 5
reduce-axis     1    2   5
reduce-axis     2    2 3 
```
or in APL-speak:
```apl
      res ← ⍳2 3 5
      ⍴res
2 3 5      
      ⍴+/[0]res 
3 5
      ⍴+/[1]res 
2 5
      ⍴+/[2]res 
2 3
```

Reply:
I left in the performant version while I study and understand how to put in a patch

Reply:
I don't see a reduction.  That is more like a form of select from what I see; there is no reduction in those versions, correct?

Reply:
I see it now.  I am implementing that function (reduce-axis) in tech.v3.tensor :-).

Reply:
Apologies, sloppy notation.  More explicitly: 
```apl 
      res ← 2 3 5 ⍴ ⍳×/2 3 5
      ⍴res
2 3 5      
      res
 0  1  2  3  4
 5  6  7  8  9
10 11 12 13 14

15 16 17 18 19
20 21 22 23 24
25 26 27 28 29
      +/[0]res # reducing along dimension 0
15 17 19 21 23
25 27 29 31 33
35 37 39 41 43
      ⍴+/[0]res # shape of reducing + along dimension 0
3 5
      +/[1]res # reducing + along dimension 1
15 18 21 24 27
60 63 66 69 72
      ⍴+/[1]res # shape of +/[1]res
2 5
      +/[2]res # reducing along dimension 2
10  35  60
85 110 135
      ⍴+/[2]res # shape of reducing along dimension 2
2 3
```

Reply:
So more generally, if the shape of a thing is `[A B C]`, and you reduce along the 0 dimension, the shape of the output should be `[B C]`.  If you reduce along dimension 1, the shape of the output should be `[A C]`.  And if you reduce along dimension 2, the output should be `[A B]`. Etc etc.

Reply:
Great example, I will make sure that works.

Reply:
https://cnuernber.github.io/dtype-next/tech.v3.tensor.html#var-reduce-axis

Reply:
Updated to use `tech.v3.tensor/reduce-axis`.  When available I'll refactor for degenerate rank 1 case.